### PR TITLE
Implement GetConnections on the cable engine.

### PR DESCRIPTION
**Pull request by mangelajo**
- [x] Unit testing

### CODE REVIEW

File: pkg/cable/ipsec/getconnections.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	client, err := getClient()
+	if err != nil {
+		return nil, err
+	}
+	defer client.Close()
+
+	sas, err := client.ListSas("", "")
+	if err != nil {
+		klog.Errorf("Error while retrieving active IKE_SAs during connection list: %v", err)
+		return nil, err
```
Comment (tpantelis) [inline pkg/cable/ipsec/getconnections.go:21]:
```suggestion
		return nil, errors.WithMessage(err, "Error while retrieving active IKE_SAs")
```

Also no need to also log here.

File: pkg/cable/ipsec/getconnections.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	return i.getSAListConnections(sas)
+}
+
+// getSAListConnections converts an IKE SA List from ListSas, and converts into a cable connection list
+// it's made a separate function for testability (unit tests)
```
Comment (tpantelis) [inline pkg/cable/ipsec/getconnections.go:28]:
```suggestion
// It's made a separate function for testability (unit tests)
```

File: pkg/cable/ipsec/getconnections.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	case "REKEYING":
+		connection.SetStatus(cable.Connected, "Connection is being re-keyed")
+
+	case "REKEYED":
+		connection.SetStatus(cable.Connected,
+			"Connection has been re-keyed, encryption alg=%s, keysize %s rekey-time=%s",
+			sa.Encr_alg, sa.Encr_keysize, sa.Rekey_time)
+
+	case "DELETING":
+		connection.SetStatus(cable.ConnectionError, "The endpoint connection is in DELETING State")
```
Comment (tpantelis) [inline pkg/cable/ipsec/getconnections.go:77]:
```suggestion
		connection.SetStatus(cable.ConnectionError, "The connection is being deleted")
```

File: pkg/cable/ipsec/getconnections.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+	case "DESTROYING":
+		connection.SetStatus(cable.ConnectionError, "The endpoint connection is in DESTROYING State")
```
Comment (tpantelis) [inline pkg/cable/ipsec/getconnections.go:80]:
```suggestion
		connection.SetStatus(cable.ConnectionError, "The connection is being destroyed")
```

I wonder what thee difference is between DELETING and DESTROYING 😄

File: pkg/cable/ipsec/getconnections.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+// updateConnectionState reads the IKE SA connection state and translates that into the cable
+// connection status, IKE SA states for strongswan can be found here:
+//  https://github.com/strongswan/strongswan/blob/d30498edf1ebda4c6f49dab8192a632b871cc7da/src/libcharon/sa/ike_sa.h#L304
+//
+func updateConnectionState(sa goStrongswanVici.IkeSa, connection *cable.Connection) {
+	switch sa.State {
+
+	case "CREATED":
+		connection.SetStatus(cable.ConnectionError, "Connection created but not started.")
```
Comment (tpantelis) [inline pkg/cable/ipsec/getconnections.go:58]:
```suggestion
		connection.SetStatus(cable.ConnectionError, "Connection has been created but not yet started")
```

File: pkg/cable/ipsec/getconnections.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+	case "CONNECTING":
+		connection.SetStatus(cable.Connecting, "Connecting to %s:%s", sa.Remote_host,
+			sa.Remote_port)
+
+	case "ESTABLISHED":
+		connection.SetStatus(cable.Connected, "Encryption alg=%s, keysize=%s rekey-time=%s",
```
Comment (tpantelis) [inline pkg/cable/ipsec/getconnections.go:65]:
```suggestion
		connection.SetStatus(cable.Connected, "Connected to %s:%s - encryption alg=%s, keysize=%s rekey-time=%s", sa.Remote_host, sa.Remote_port
```

It  seems useful to include the host:port same as for CONNECTING?  In fact should we include host:port in all the other status messages, ie "Connection to %s:%s is ..."?

File: pkg/cable/ipsec/getconnections.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+			sa.Encr_alg, sa.Encr_keysize, sa.Rekey_time)
+
+	case "DELETING":
+		connection.SetStatus(cable.ConnectionError, "The endpoint connection is in DELETING State")
+
+	case "DESTROYING":
+		connection.SetStatus(cable.ConnectionError, "The endpoint connection is in DESTROYING State")
+
+	case "PASSIVE":
+		connection.SetStatus(cable.ConnectionError, "Connection is in passive state.")
```
Comment (tpantelis) [inline pkg/cable/ipsec/getconnections.go:83]:
```suggestion
		connection.SetStatus(cable.ConnectionError, "Connection is in a passive state")
```

File: pkg/cable/ipsec/strongswan.go [hunk 8]
```diff
@@ -203,6 +204,8 @@ func (i *strongSwan) ConnectToEndpoint(endpoint types.SubmarinerEndpoint) (strin
 		return "", fmt.Errorf("error loading connection %q with config %#v into charon: %v", endpoint.Spec.CableName, ikeConf, err)
 	}
 
+	i.remoteEndpoints[endpoint.Spec.CableName] = endpoint.Spec
```
Comment (tpantelis) [inline pkg/cable/ipsec/strongswan.go:207]: I guess we're safe to assume thread safety here - we know the lone caller locks.

Comment (mangelajo) [reply pkg/cable/ipsec/strongswan.go:207]: Yes, that's the assumption, we could add a lock to the cable methods if it changed.

File: pkg/cable/ipsec/getconnections.go (hunk 1 above)
Comment (mangelajo) [reply pkg/cable/ipsec/getconnections.go:21]: Nice I didn't know that errros.WithMessage :-)

File: pkg/cable/ipsec/getconnections.go (hunk 2 above)
Comment (mangelajo) [reply pkg/cable/ipsec/getconnections.go:28]: Done.

File: pkg/cable/ipsec/getconnections.go (hunk 5 above)
Comment (mangelajo) [reply pkg/cable/ipsec/getconnections.go:58]: Done.

File: pkg/cable/ipsec/getconnections.go (hunk 3 above)
Comment (mangelajo) [reply pkg/cable/ipsec/getconnections.go:77]: Done.

File: pkg/cable/ipsec/getconnections.go (hunk 4 above)
Comment (mangelajo) [reply pkg/cable/ipsec/getconnections.go:80]: Yeah, me too, I didn't want to dig :hole:

Comment (mangelajo) [reply pkg/cable/ipsec/getconnections.go:80]: Done.

File: pkg/cable/ipsec/getconnections.go (hunk 7 above)
Comment (mangelajo) [reply pkg/cable/ipsec/getconnections.go:83]: Done

File: pkg/cable/ipsec/getconnections.go (hunk 6 above)
Comment (mangelajo) [reply pkg/cable/ipsec/getconnections.go]: Oh this one is missing, doing and force-pushing.

Comment (mangelajo) [reply pkg/cable/ipsec/getconnections.go]: Done

**Comment by mangelajo** [comment]
Im re-checking, because E2E passes locally on my computer :?

**Comment by mangelajo** [comment]
I was crashing the cableEngine due to an uninitialized map, fixed on last commit.

**Pull request merged**